README.md
# Django TailorDev Biblio

Bibliography management with Django.

[![](https://travis-ci.org/TailorDev/django-tailordev-biblio.svg?branch=master)
](https://travis-ci.org/TailorDev/django-tailordev-biblio/)
[![Coverage Status](https://coveralls.io/repos/github/TailorDev/django-tailordev-biblio/badge.svg?branch=master)](https://coveralls.io/github/TailorDev/django-tailordev-biblio?branch=master)
[![](https://img.shields.io/pypi/v/django-tailordev-biblio.svg)](https://pypi.python.org/pypi/django-tailordev-biblio)

## Compatibility

Since the `1.0.0` release, we have added full support for recent python and
Django releases:

|            | Django 1.8         | Django 1.9         | Django 1.10        | Django 1.11        | Django 2.0         | Django 2.1         |
| ---------- | ------------------ | ------------------ | ------------------ | ------------------ | ------------------ | ------------------ |
| Python 2.7 | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |                    |                    |
| Python 3.4 | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |                    |
| Python 3.5 | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |
| Python 3.6 | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |

Please note that for older Python and/or Django versions, you can still use the `0.3` release.

## Sandbox

Click on the screenshot below to browse the sandbox deployed on Heroku:

[
![Django TailorDev Biblio Screenshot](docs/images/preview.png "Checkout the demo!")
](https://tailordev-biblio.herokuapp.com)

_Nota bene:_ to test [references
importation](https://tailordev-biblio.herokuapp.com/import/) from PMIDs or DOIs,
you will need to login with these credentials:

```
login: rosalind
password: idiscovereddnastructurefirst
```

## Installation

### Install `td_biblio`

The easiest way to go is to use pip:

```bash
$ pip install -U django-tailordev-biblio
```

### Configure `td_biblio`

Add `td_biblio` to your `INSTALLED_APPS` in django settings:

```python
# foo_project/settings.py

INSTALLED_APPS = (
    # other apps…
    'td_biblio',
)
```

Add `td_biblio` urls your project url patterns:

```python
# foo_project/urls.py

urlpatterns = [
    # other urls…
    url(r'^bibliography/', include('td_biblio.urls', namespace='td_biblio')),
]
```

And finally migrate your database from your project root path:

```bash
$ python manage.py migrate td_biblio
```

### Add a base template

In order to use `td_biblio` templates, you will need to create a base template
to inherit from. This base template should be visible as `_layouts/base.html`
and contains at least the following blocks:

```html
<html>
  <head>
    <title>Publication list</title>
  </head>
  <body>
    {% block content %}{% endblock content %} {% block javascripts %}{% endblock
    javascripts %}
  </body>
</html>
```

As you might have guessed, the `content` block is the base block where we render
the bibliography list and item details, while the `javascripts` block contains
eponym front-end dependencies. You will find an example base layout template at:
[`td_biblio/templates/_layouts/base.html`](https://github.com/TailorDev/django-tailordev-biblio/blob/master/td_biblio/templates/_layouts/base.html)

## Import bibliography

Once `td_biblio` is installed and configured, you may want to import your
references stored in a BibTeX file. Hopefully, there is a command for that:

```bash
$ python manage.py bibtex_import my_bibliography.bib
```

Alternatively, you can browse to the `/import/` view with a superuser account
to fetch references from their DOIs or PMIDs.

## Hack

### Development installation

If you intend to work on the code, clone this repository and install all
dependencies in a virtual environment via:

```bash
$ make bootstrap
```

And then start the development server via:

```bash
$ make dev
```

### Running the Tests

You can run the tests with via:

```bash
$ make test
```

### Deploy to Heroku

To start playing with an Heroku instance, we suppose you have:

- an Heroku account;
- a functional SSH key imported for this account;
- installed the [`heroku` client](https://devcenter.heroku.com/articles/heroku-cli).

```bash
# Log in to heroku
$ heroku login

# Configure current repository as a registered heroku app
$ heroku git:remote -a tailordev-biblio

# Push your code
$ git push heroku master

# Perform database migrations
$ heroku run python sandbox/manage.py migrate
```

### Upload a new release to PyPI

```bash
$ python setup.py sdist bdist_wheel
$ twine upload dist/*
```

## License

`django-tailordev-biblio` is released under the MIT License. See the bundled
LICENSE file for details.


CHANGELOG.md
# TailorDev Biblio

## 2.0.0 (November 13, 2018)

- Add Django 2+ compatibility
- Add authors duplication cleanup form (#22)

### BC Breaks

- Drops compatibility with Django 1.7

## 1.2.0 (December 1, 2017)

- Make PMIDLoader & DOILoader bullet-proof

## 1.1.0 (June 14, 2017)

- Add PMIDLoader & DOILoader
- Add references importation view for admins (from PMIDs or DOIs)
- Better integrate the sandbox
- Make the sandbox deployable to Heroku

## 1.0.0 (March 23, 2017)

- Add support for python 3.4+
- Add support for Django 1.7+
- Switch to `py.test` test runner
- Remove jQuery implicit dependency

### BC Breaks

- Drops compatibility with python 2.6
- Drops compatibility with Django < 1.7

### 1.0.1 (March 23, 2017)

- Fix missing dev requirments

## 0.3 (February 3, 2015)

- Add publication list partial template

## 0.2 (June 2, 2014)

- Add support for partial publication date

### 0.2.1 (February 3, 2015)

- Entry first and last author are now object properties

## 0.1 (January 10, 2014)

First public release. Main features are:

- basic bibliography management
- bibtex file import
- django users/authors linking
- reference list view filtering by author, date, collections

TODO:

- Work on the documentation
- Add EndNote support


sandbox/settings.py
"""
Django settings for td_biblio sandbox.
"""

import os

import dj_database_url

try:
    from django.core.urlresolvers import reverse_lazy
except:
    from django.urls import reverse_lazy

# Build paths inside the project like this: os.path.join(BASE_DIR, ...)
BASE_DIR = os.path.dirname(os.path.abspath(__file__))

# Default secret key is for development only!
SECRET_KEY = os.environ.get(
    "DJANGO_SECRET_KEY", "2#jyyh#-tgk$%wdx0zgfx=-)#hr9ni(_m66a3!rmco8^a5cns!"
)

# SECURITY WARNING: don't run with debug turned on in production!
DEBUG = os.environ.get("IS_PUBLIC_SANDBOX") is None

ALLOWED_HOSTS = ["localhost", "127.0.0.1", "tailordev-biblio.herokuapp.com"]

LOGIN_URL = reverse_lazy("login")
LOGIN_REDIRECT_URL = reverse_lazy("td_biblio:import")

# Application definition
INSTALLED_APPS = [
    "django.contrib.admin",
    "django.contrib.auth",
    "django.contrib.contenttypes",
    "django.contrib.sessions",
    "django.contrib.messages",
    "whitenoise.runserver_nostatic",
    "django.contrib.staticfiles",
    "td_biblio",
]

MIDDLEWARE = (
    "django.middleware.security.SecurityMiddleware",
    "whitenoise.middleware.WhiteNoiseMiddleware",
    "django.contrib.sessions.middleware.SessionMiddleware",
    "django.middleware.common.CommonMiddleware",
    "django.middleware.csrf.CsrfViewMiddleware",
    "django.contrib.auth.middleware.AuthenticationMiddleware",
    "django.contrib.messages.middleware.MessageMiddleware",
    "django.middleware.clickjacking.XFrameOptionsMiddleware",
)

# Django < 1.10 compat
MIDDLEWARE_CLASSES = (
    "django.contrib.sessions.middleware.SessionMiddleware",
    "django.middleware.common.CommonMiddleware",
    "django.middleware.csrf.CsrfViewMiddleware",
    "django.contrib.auth.middleware.AuthenticationMiddleware",
    "django.contrib.auth.middleware.SessionAuthenticationMiddleware",
    "django.contrib.messages.middleware.MessageMiddleware",
    "django.middleware.clickjacking.XFrameOptionsMiddleware",
)

ROOT_URLCONF = "sandbox.urls"

TEMPLATES = [
    {
        "BACKEND": "django.template.backends.django.DjangoTemplates",
        "DIRS": [os.path.join(BASE_DIR, "templates")],
        "APP_DIRS": True,
        "OPTIONS": {
            "context_processors": [
                "django.template.context_processors.debug",
                "django.template.context_processors.request",
                "django.contrib.auth.context_processors.auth",
                "django.contrib.messages.context_processors.messages",
            ]
        },
    }
]

# Django < 1.8 compat
TEMPLATE_DIRS = [os.path.join(BASE_DIR, "templates")]

WSGI_APPLICATION = "sandbox.wsgi.application"


# Database
DATABASES = {"default": dj_database_url.config(conn_max_age=600)}


# Password validation
AUTH_PASSWORD_VALIDATORS = [
    {
        "NAME": "django.contrib.auth.password_validation.UserAttributeSimilarityValidator"  # noqa
    },
    {"NAME": "django.contrib.auth.password_validation.MinimumLengthValidator"},  # noqa
    {"NAME": "django.contrib.auth.password_validation.CommonPasswordValidator"},  # noqa
    {
        "NAME": "django.contrib.auth.password_validation.NumericPasswordValidator"  # noqa
    },
]


# Internationalization
LANGUAGE_CODE = "en-us"
TIME_ZONE = "UTC"
USE_I18N = True
USE_L10N = True
USE_TZ = True


# Static files (CSS, JavaScript, Images)
STATIC_URL = "/static/"
STATIC_ROOT = os.path.join(BASE_DIR, "public", "static")
STATICFILES_DIRS = (os.path.join(BASE_DIR, "static"),)
